Document/0x06h-Testing-Platform-Interaction.md

```diff
@@ -3,365 +3,2956 @@
+- the scheme of the `webpageURL` must be HTTP or HTTPS (any other scheme should throw an exception). The [`scheme` instance property](https://developer.apple.com/documentation/foundation/urlcomponents/1779624-scheme "URLComponents scheme") of `URLComponents` / `NSURLComponents` can be used to verify this.
+
+If you don't have the original source code you can use Radare2 or Rabin2 to search the binary strings for the link receiver method:
+
+```bash
+$ rabin2 -zq Telegram\ X.app/Telegram\ X | grep restorationHan
+
+0x1000deea9 53 52 application:continueUserActivity:restorationHandler:
+```
+
+###### Checking the Data Handler Method
+
+You should check how the received data is validated. Apple [explicitly warns about this](https://developer.apple.com/documentation/uikit/core_app/allowing_apps_and_websites_to_link_to_your_content/handling_universal_links "Handling Universal Links"):
+
+> Universal links offer a potential attack vector into your app, so make sure to validate all URL parameters and discard any malformed URLs. In addition, limit the available actions to those that do not risk the user’s data. For example, do not allow universal links to directly delete content or access sensitive information about the user. When testing your URL-handling code, make sure your test cases include improperly formatted URLs.
+
+As stated in the [Apple Developer Documentation](https://developer.apple.com/documentation/uikit/core_app/allowing_apps_and_websites_to_link_to_your_content/handling_universal_links "Handling Universal Links"), when iOS opens an app as the result of a universal link, the app receives an `NSUserActivity` object with an `activityType` value of `NSUserActivityTypeBrowsingWeb`. The activity object’s `webpageURL` property contains the HTTP or HTTPS URL that the user accesses. The following example in Swift from the Telegram app verifies exactly this before opening the URL:
+
+```swift
+func application(_ application: UIApplication, continue userActivity: NSUserActivity, 
+                restorationHandler: @escaping ([Any]?) -> Void) -> Bool {
+    ...
+    if userActivity.activityType == NSUserActivityTypeBrowsingWeb, let url = userActivity.webpageURL {
+        self.openUrl(url: url)
+    }
+    
+    return true
+}
+```
+
+However, depending on the app, it would be a good idea to additionally apply a URL whitelist at this point and, if the URL includes parameters, ensure that they are properly sanitized and all malformed URLs are discarded. The `NSURLComponents` API can be used to parse and manipulate the components of the URL. This can be also part of the method `application:continueUserActivity:restorationHandler:` itself or might occur on a separate method being called from it. The following [example](https://developer.apple.com/documentation/uikit/core_app/allowing_apps_and_websites_to_link_to_your_content/handling_universal_links#3001935 "An example of handling a universal link") demonstrates this:
```

Comment:
I would add that without whitelisting, none of the received data can be trusted and should be treated as such. Not only the URL needs to be "sanitised", but it should be noted that any data can be spoofed by an attacking application.

Reply:
I've considered this now in the text. I've kept the whitelist but to highlight that even if one is present the data cannot be trusted anyway. Please, let me know what you think.

Reply:
Better now :)